src/ray/common/id.h

```diff
@@ -393,10 +394,15 @@ T BaseID<T>::FromRandom() {
 
 template <typename T>
 T BaseID<T>::FromBinary(const std::string &binary) {
-  RAY_CHECK(binary.size() == T::Size())
-      << "expected size is " << T::Size() << ", but got " << binary.size();
+  return FromBinary(binary.data(), binary.size());
+}
+
+template <typename T>
+T BaseID<T>::FromBinary(const void *data, size_t size) {
+  RAY_CHECK(data != nullptr && size == T::Size())
```

Comment:
> I think it's too dangerous to do this.
> 
> You can't make sure others pass the correct size of data, right?
> emm, like what if I write this code?
> 
> ```
> string s = "123";  // actually the size of s is 3.
> auto id = XxxID::FromBinary(s.data(), XxxID::Size());
> ```
> 
> It might touch some unknown memory when you do `memcpy` below.

Yes, it does produce unexpected results if the user passes a mismatched data and size, but it does reduce the cost of copying.

e.g.
`std::string` has the constructor `string(const char * s, size_type n)`, and user can still write like this:
```
std::string a = "123";
std::string b(a.data(), 40); // actually size is 3
```
We could not say that `std::string` should not provide such constructor.

Reply:
But the usage you wrote in this PR pass a wrong size in:
```C++
TaskID TaskID::ComputeDriverTaskId(const WorkerID &driver_id) {
  return TaskID::FromBinary(driver_id.Data(), TaskID::Size());
}
```
Currently the size of `driver_id` and `TaskID::Size()` are not equal.
Image what'll happen if size of `driver_id` is shorter than `TaskID::Size()`.


https://github.com/ray-project/ray/blob/1cd567d5ed3744cffaf0d6583b0a5b9319cdf31a/src/ray/common/id.cc#L276to#L278

Reply:
> But the usage you wrote in this PR passes a wrong size in:
> 
> ```c++
> TaskID TaskID::ComputeDriverTaskId(const WorkerID &driver_id) {
>   return TaskID::FromBinary(driver_id.Data(), TaskID::Size());
> }
> ```
> 
> Currently the size of `driver_id` and `TaskID::Size()` are not equal.
> Image what'll happen if size of `driver_id` is shorter than `TaskID::Size()`.
> 
> https://github.com/ray-project/ray/blob/1cd567d5ed3744cffaf0d6583b0a5b9319cdf31a/src/ray/common/id.cc#L276to#L278

Yes, my bad. I ignore that the size of WorkerID less than TaskID. I will fix it soon.